src/containers/useForm.ts

```diff
@@ -0,0 +1,208 @@
+ *  -> isSubmitDisabled - is submit button disabled?
+ *
+ * initForm and form objects hold whole state of a form.
+ * their structure:
+ *  -> [key]:   -- input field ID
+ *    -> value  -- field value
+ *    -> errorMessage -- actual error messages (in case of an error)
+ *    -> state  -- state of a field ('default', 'success', 'error')
+ *    -> validation:    -- means of validation
+ *      -> isRequired   -- is field required?
+ *      -> validators:  -- validation functions and their error messages
+ *        -> validator    -- validation function
+ *        -> errorMessage -- error message that should be set in a case of an error
+ */
+export const useForm = (initFields: Omit<Omit<IFields, 'errorMessages'>, 'state'>, callback: Function) => {
+  const transformFormData = useCallback(
+    (values?: IFieldValues): IFields => {
+      let defaultFields: IFields = {};
+      for (const key in initFields) {
+        defaultFields[key] = {};
+
+        // init data
+        defaultFields[key].isRequired = initFields[key].isRequired;
+        // SHALLOW COPY (should not change anyway)
+        defaultFields[key].validators = initFields[key].validators;
+
+        // additional data
+        defaultFields[key].state = 'default';
+        if (values?.[key]) {
+          defaultFields[key].value = values[key];
+        } else if (initFields[key].value) {
+          defaultFields[key].value = initFields[key].value;
+        } else {
+          defaultFields[key].value = '';
+        }
+      }
+      return defaultFields;
+    },
+    [initFields]
+  );
+
+  const [fields, setFields] = useState<IFields>(transformFormData());
+
+  // is submit button disabled?
+  const [isSubmitDisabled, setIsSubmitDisabled] = useState<boolean>(true);
+  // has any field been changed?
+  // important for edit page (do not submit until any new content)
+  const [hasChanged, setHasChanged] = useState<boolean>(false);
+
+  // are all validated inputs valid?
+  const isFormValid = useCallback(() => {
+    for (const key in fields) {
+      if (fields[key].errorMessages?.length) return false;
+    }
+
+    return true;
+  }, [fields]);
+
+  // are all required inputs filled?
+  const areRequiredFilled = useCallback(() => {
+    for (const key in fields) {
+      if (fields[key].isRequired && !fields[key].value) {
+        return false;
+      }
+    }
+
+    return true;
+  }, [fields]);
+
+  // callback (on change of an input)
+  const onChange = (fieldName: any, fieldValue: any) => {
+    // also delete old error messages, new checks are going to be done
+    const newFieldState = {
+      ...fields[fieldName],
+      value: fieldValue,
+      errorMessages: [],
+      state: 'default' as TextInputProps['validated'],
+    };
+    validate(newFieldState);
+    setFields({ ...fields, [fieldName]: newFieldState });
+
+    setHasChanged(true);
+  };
+
+  // validate field state and change error messages / state
+  const validate = (fieldState: IField) => {
```

Comment:
let's use only `field` instead of `fieldState` everywhere

Even if `addError` does the check for empty error, I would call this method only if you know you want to add new error

```js
const validate = (fieldState: IField) => {
  if (fieldState.isRequired) {
    if (!fieldState.value) {
      addError(fieldState, 'Field must be filled.');
    }
    setState(fieldState);
  }
  if (fieldState.validators) {
    for (const validator of fieldState.validators) {
      if (!validator.validator(fieldState.value)) {
        addError(fieldState, validator.errorMessage);
      }
    }
    setState(fieldState);
  }
};
```